File: Software/Java/src/com/dexterind/gopigo/behaviours/Motion.java
/*
 * #%L
 * **********************************************************************
 * ORGANIZATION  :  DexterIndustries
 * PROJECT       :  GoPiGo Java Library
 * FILENAME      :  Motion.java
 * AUTHOR        :  Marcello Barile <marcello.barile@gmail.com>
 *
 * This file is part of the GoPiGo Java Library project. More information about
 * this project can be found here:  https://github.com/DexterInd/GoPiGo
 * **********************************************************************
 * %%
 * GoPiGo for the Raspberry Pi: an open source robotics platform for the Raspberry Pi.
 * Copyright (C) 2017  Dexter Industries

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/gpl-3.0.txt>.
 *
 * #L%
 */
package com.dexterind.gopigo.behaviours;

import java.io.IOException;
import java.util.*;

import com.dexterind.gopigo.components.*;
import com.dexterind.gopigo.utils.*;

/**
 * Handles all the methods related to the movement of the GoPigo.
 * @author marcello
 *
 */
public class Motion {
  /**
   * The main object which handles the methods to get access to the resources
   * connected to the board.
   */
  private static Board board;
  /**
   * The instance of the current object.
   */
  private static Motion instance = null;
  /**
   * The HashMap of the directions. It could be useful as helper for rotating the
   * servo motor or the GoPiGo itself.
   */
  public Map<String, Integer> directions = new HashMap<String, Integer>();
  /**
   * The debug object.
   */
  private Debug debug;

  public Motion() throws IOException, InterruptedException {
    directions.put("n", 0);
    directions.put("ne", 45);
    directions.put("e", 90);
    directions.put("se", 135);
    directions.put("s", 180);
    directions.put("sw", 225);
    directions.put("w", 270);
    directions.put("nw", 315);

    board = Board.getInstance();
  }

  /**
   * Provides a global point of access to the Motion instance.
   * @return  the <code>Motion</code> instance.
   */
  public static Motion getInstance() throws IOException, InterruptedException {
    if(instance == null) {
      instance = new Motion();
    }
    return instance;
  }

  /**
   * Moves forward the Gopigo.
   * @param usePid A boolean flag which defines if it's needed to use or not the PID.
   * @return A status code.
   * @throws IOException
   */
  public int forward(boolean usePid) throws IOException {
    int command = usePid ? Commands.FWD : Commands.MOTOR_FWD;
    return board.writeI2c(command, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Moves backward the Gopigo.
   * @param usePid A boolean flag which defines if it's needed to use or not the PID.
   * @return A status code.
   * @throws IOException
   */
  public int backward(boolean usePid) throws IOException {
    int command = usePid ? Commands.BWD : Commands.MOTOR_BWD;
    return board.writeI2c(command, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Moves the Gopigo to the left.
   * @return A status code.
   * @throws IOException
   */
  public int left() throws IOException {
    return board.writeI2c(Commands.LEFT, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Rotates the Gopigo to the left.
   * @return A status code.
   * @throws IOException
   */
  public int leftWithRotation() throws IOException {
    return board.writeI2c(Commands.LEFT_ROT, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   *  Moves the Gopigo to the right.
   * @return A status code.
   * @throws IOException
   */
  public int right() throws IOException {
    return board.writeI2c(Commands.RIGHT, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Rotates the Gopigo to the right.
   *
   * @return A status code.
   * @throws IOException
   */
  public int rightWithRotation() throws IOException {
    return board.writeI2c(Commands.RIGHT_ROT, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Stops any kind of movement.
   * @return A status code.
   * @throws IOException
   */
  public int stop() throws IOException {
    return board.writeI2c(Commands.STOP, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Increases the speed of both the motors.
   *
   * @return A status code.
   * @throws IOException
   */
  public int increaseSpeed() throws IOException {
    return board.writeI2c(Commands.ISPD, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Decreases the speed of both the motors.
   * @return A status code.
   * @throws IOException
   */
  public int decreaseSpeed() throws IOException {
    return board.writeI2c(Commands.DSPD, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Sets the speed only for the left motor.
   * @param speed A speed value between 0 and 255
   * @return A status code.
   * @throws IOException
   */
  public int setLeftSpeed(int speed) throws IOException {
    if (speed > 255) {
      speed = 255;
    } else if (speed < 0) {
      speed = 0;
    }

    return board.writeI2c(Commands.SET_LEFT_SPEED, speed, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Sets the speed only for the right motor.
   * @param speed A speed value between 0 and 255.
   * @return A status code.
   * @throws IOException
   */
  public int setRightSpeed(int speed) throws IOException {
    if (speed > 255) {
      speed = 255;
    } else if (speed < 0) {
      speed = 0;
    }

    return board.writeI2c(Commands.SET_RIGHT_SPEED, speed, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Sets the speed for both the motors.
   * @param speed A speed value between 0 and 255
   * @return A status code.
   * @throws IOException
   */
  public int setSpeed(int speed) throws IOException {
    if (speed > 255) {
      speed = 255;
    } else if (speed < 0) {
      speed = 0;
    }

    setLeftSpeed(speed);
    setRightSpeed(speed);

    return Statuses.OK;
  }

  /**
   * Executes a Trim test.
   * @param value The value to test between -100 and 100.
   * @return A status code.
   * @throws IOException
   */
  public double trimTest(int value) throws IOException {
    if (value > 100) {
      value = 100;
    } else if (value < -100) {
      value = -100;
    }
    value +=100;

    return board.writeI2c(Commands.TRIM_TEST, value, Commands.UNUSED, Commands.UNUSED);
  }

  /**
   * Reads the current Trim value.
   * @return The current Trim value between -100 and 100.
   * @throws IOException
   */
  public double trimRead() throws IOException {
    board.writeI2c(Commands.TRIM_READ, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
    board.sleep(80);

    byte[] b1 = board.readI2c(1);
    byte[] b2 = board.readI2c(1);
    int val1 = (int)b1[0] & 0xFF;
    int val2 = (int)b2[0] & 0xFF;

    if (val1 != -1 && val2 != -1) {
      int v = val1 * 256 + val2;
      if (v == 255) {
        return -3;
      }
      return v;
    } else {
      return Statuses.ERROR;
    }
  }

  /**
   * Sets the Trim value.
   * @param value The Trim value between -100 and 100.
   * @return A status code.
   * @throws IOException
   */
  public double trimWrite(int value) throws IOException {
    if (value > 100) {
      value = 100;
    } else if (value < -100) {
      value = -100;
    }
    value +=100;

    return board.writeI2c(Commands.TRIM_WRITE, value, Commands.UNUSED, Commands.UNUSED);
  }

 /**
  * Reads the timeout status.
  * @return The timeout value.
  * @throws IOException
  */
  public int readTimeoutStatus() throws IOException {
    int[] status = board.readStatus();
    return status[1];
  }

  /**
   * Enables the communication timeout.
   * @param timeout The value of the timout.
   * @return A status code.
   * @throws IOException
   */
  public int enableCommunicationTimeout(int timeout) throws IOException {
    return board.writeI2c(Commands.EN_COM_TIMEOUT, timeout / 256, timeout % 256, Commands.UNUSED);
  }

  /**
   * Disables the communication timeout.
   * @return A status code.
   * @throws IOException
   */
  public int disableCommunicationTimeout() throws IOException {
    return board.writeI2c(Commands.DIS_COM_TIMEOUT, Commands.UNUSED, Commands.UNUSED, Commands.UNUSED);
  }
}

File: Software/NodeJS/libs/behaviours/motion.js
var commands   = require('../commands')

Motion.directions = {
  'n' : 0,
  'ne': 45,
  'e' : 90,
  'se': 135,
  's' : 180,
  'sw': 225,
  'w' : 270,
  'nw': 315
}

var self

function Motion(gopigo) {
  self = this
  this.gopigo = gopigo
}

Motion.prototype.forward = function(usePid) {
  var command = usePid ? commands.fwd : commands.motor_fwd
  return this.gopigo.board.writeBytes(command.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.backward = function(usePid) {
  var command = usePid ? commands.bwd : commands.motor_bwd
  return this.gopigo.board.writeBytes(command.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.left = function() {
  return this.gopigo.board.writeBytes(commands.left.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.leftWithRotation = function() {
  return this.gopigo.board.writeBytes(commands.left_rot.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.right = function() {
  return this.gopigo.board.writeBytes(commands.right.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.rightWithRotation = function() {
  return this.gopigo.board.writeBytes(commands.right_rot.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.stop = function() {
  return this.gopigo.board.writeBytes(commands.stop.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.increaseSpeed = function() {
  return this.gopigo.board.writeBytes(commands.ispd.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.decreaseSpeed = function() {
  return this.gopigo.board.writeBytes(commands.dspd.concat([commands.unused, commands.unused, commands.unused]))
}
Motion.prototype.setLeftSpeed = function(speed) {
  if (speed > 255)
    speed = 255
  else if (speed < 0)
    speed = 0

  return this.gopigo.board.writeBytes(commands.set_left_speed.concat([speed, commands.unused, commands.unused]))
}
Motion.prototype.setRightSpeed = function(speed) {
  if (speed > 255)
    speed = 255
  else if (speed < 0)
    speed = 0

  return this.gopigo.board.writeBytes(commands.set_right_speed.concat([speed, commands.unused, commands.unused]))
}
Motion.prototype.setSpeed = function(speed) {
  this.setLeftSpeed(speed)
  this.setRightSpeed(speed)
}
Motion.prototype.trimTest = function(value) {
  if (value > 100)
    value = 100
  else if (value < -100)
    value = -100
  value += 100

  return this.gopigo.board.writeBytes(commands.trim_test.concat([value, commands.unused, commands.unused]))
}
Motion.prototype.trimRead = function() {
  var write = this.gopigo.board.writeBytes(commands.trim_read.concat([this.pin, commands.unused, commands.unused]))
  if (write) {
    this.gopigo.board.wait(80)
    var b1 = this.gopigo.board.readByte()
    var b2 = this.gopigo.board.readByte()
    if (b1[0] != -1 && b2[0] != -1) {
      var val = b1[0] * 256 + b2[0]
      if (val == 255)
        return -3
      return val
    } else
      return false
  } else {
    return false
  }
}
Motion.prototype.trimWrite = function(value) {
  if (value > 100)
    value = 100
  else if (value < -100)
    value = -100
  value += 100

  return this.gopigo.board.writeBytes(commands.trim_write.concat([value, commands.unused, commands.unused]))
}
Motion.prototype.readTimeoutStatus = function() {
  var status = this.gopigo.board.readStatus()
  return status[1]
}
Motion.prototype.enableCommunicationTimeout = function(timeout) {
  return this.gopigo.board.writeBytes(commands.en_com_timeout.concat([timeout, commands.unused, commands.unused]))
}
Motion.prototype.disableCommunicationTimeout = function() {
  return this.gopigo.board.writeBytes(commands.dis_com_timeout.concat([commands.unused, commands.unused, commands.unused]))
}

module.exports = Motion